### PR TITLE
Podman V2 compatibility fixes

### PR DESCRIPTION
**Pull request by HarryMichal**
Podman is nearing a new major version (V2). We need to preserve compatibility.

### CODE REVIEW

File: src/cmd/create.go [hunk 1]
```diff
@@ -338,7 +338,6 @@ func createContainer(container, image, release string, showCommandToEnter bool)
 		"create",
 		"--dns", "none",
 		"--env", toolboxPathEnvArg,
-		"--group-add", sudoGroup,
```
Comment (debarshiray) [inline src/cmd/create.go:341]: Do we really not need the group for `sudo`?

File: src/cmd/create.go [hunk 2]
```diff
@@ -347,14 +346,13 @@ func createContainer(container, image, release string, showCommandToEnter bool)
...
 		"--pid", "host",
-		"--privileged",
 		"--security-opt", "label=disable",
 	}
 
 	createArgs = append(createArgs, ulimitHost...)
 
 	createArgs = append(createArgs, []string{
-		"--userns=keep-id",
+		// "--userns=keep-id",
```
Comment (debarshiray) [inline src/cmd/create.go:355]: This cannot be right. :)

Wasn't there a bug in Podman that broke `--userns=keep-id` for rootless containers that later got fixed?

File: src/cmd/list.go [hunk 3]
```diff
@@ -211,12 +216,17 @@ func listOutput(images, containers []map[string]interface{}) {
...
+		var id, name, created, status, imageName string
 		for _, container := range containers {
-			id := container["ID"].(string)[:12]
-			name := container["Names"].(string)
-			created := container["Created"].(string)
-			status := container["Status"].(string)
-			imageName := container["Image"].(string)
+			id = container["Id"].(string)[:12]
+			if podman.CheckVersion("2.0.0") {
+				name = container["Names"].([]interface{})[0].(string)
```
Comment (debarshiray) [inline src/cmd/list.go:223]: Ugh. Another JSON change. :/

I wonder if this is needed in other places.

File: src/cmd/create.go (hunk 2 above)
Comment (HarryMichal) [reply src/cmd/create.go:355]: Yes, exactly. I was playing with this and pushed it with it being commented. It should be uncommented, of course.

File: src/cmd/create.go (hunk 1 above)
Comment (HarryMichal) [reply src/cmd/create.go:341]: Well, in the case of Fedora-based images it works but other distros may use `wheel` as the sudoers group and this can't be tested on the host but only inside of the container.

File: src/cmd/list.go (hunk 3 above)
Comment (HarryMichal) [reply src/cmd/list.go:223]: I have some new code in https://github.com/containers/toolbox/pull/469 that makes this a bit less painful. But yeah, we need to be careful about this.

**Comment by HarryMichal** [comment]
This relies on #469 so that the number of version checks against Podman is kept to minimum.

File: src/cmd/create.go (hunk 1 above)
Comment (debarshiray) [reply src/cmd/create.go]: For those following along at home, this was because of a [regression](https://github.com/containers/libpod/pull/6715) in Podman where the `--group-add` and `--privileged` flags won't work together.

File: src/cmd/create.go (hunk 2 above)
Comment (debarshiray) [reply src/cmd/create.go]: For those following along at home, I was referring to https://github.com/containers/libpod/issues/6652

File: src/cmd/list.go [hunk 4]
```diff
@@ -211,12 +216,29 @@ func listOutput(images, containers []map[string]interface{}) {
...
+		}
+
+		var id, name, created, status, imageName interface{}
 		for _, container := range containers {
-			id := container["ID"].(string)[:12]
-			name := container["Names"].(string)
-			created := container["Created"].(string)
-			status := container["Status"].(string)
-			imageName := container["Image"].(string)
+			id = utils.ShortID(container[idKey].(string))
```
Comment (debarshiray) [inline src/cmd/list.go:233]: Nitpick: this bit should be split out into a separate commit since it's unrelated to Podman 2.0.

See: https://github.com/containers/toolbox/pull/479

Comment (HarryMichal) [reply src/cmd/list.go:233]: Ah, right. Thanks for the catch.

File: src/cmd/list.go [hunk 5]
```diff
@@ -187,10 +187,15 @@ func listOutput(images, containers []map[string]interface{}) {
...
+
+		if podman.CheckVersion("2.0.0") {
+			idKey = "Id"
+		}
+
 		for _, image := range images {
 			id := utils.ShortID(image[idKey].(string))
-			name := image[nameKey].([]interface{})[0].(string)
-			created := image[createdKey].(string)
+			name := image[nameKey].([]interface{})[0]
```
Comment (debarshiray) [inline src/cmd/list.go:197]: After pondering over this a bit, I don't think we should drop the type cast just like that.

I understand that the strict cast makes the code more sensitive to (future) changes in the JSON, and currently it doesn't matter because everything is a `string` anyway. However, we should also be in control of what we are printing, and we shouldn't just throw some unknown type at the `%s` format specifier in the hope that it will somehow work out.

Go does offer some nice features like [type assertions](https://tour.golang.org/methods/15) and [type switches](https://tour.golang.org/methods/16). These could be used to come up with a more elegant solution that's more tolerant towards JSON changes, and at the same time lets us retain control over our types.

At the very least, if something is not of the expected type, we should log some debug message with details about the actual type. Parsing the JSON and dealing with all the `interface` types can be tricky at times, so logging the actual failure can be useful.

As for the actual output, we could either print an empty string for all values that aren't of the expected type, or we can skip the entire column altogether. Or maybe something else. It somewhat depends on how practical each of these options are.

Either way, we should use a separate pull request for this topic because it's not directly connected to the issue of *JSON changes in Podman 2.0*. :)

Comment (HarryMichal) [reply src/cmd/list.go]: Sure! This just looked to me like the easiest solution if something were to change and we weren't fast enough to change it. Using type assertions/switches definitely gives us more power over dealing with what to show/hide in case of an error.

I'll look into that in the future.
